src/net_processing.cpp

```diff
@@ -3206,6 +3206,10 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             return;
         }
 
+        // Stop processing the transaction early if we are still in IBD since we don't
+        // have enough information to validate it yet. Not a protocol violation.
```

Comment:
This comment is confusing to me -- have we decided whether it's protocol-violating to not process unrequested transactions?  That was exactly what I proposed we do elsewhere, in general, but there was debate over whether such a change would be breaking or whether it was useful.

If it's not useful for our software to do this in general, I don't know why it's useful to do in this smaller specific case?  And on the flip side, if it's not a protocol violation here, why exactly would it be a protocol violation to do this in general?

Note also that being in IBD happens not just when we're syncing the whole blockchain, but whenever a node is started up again after being off for a few days.  So I believe that if there are any software projects that send unrequested transactions at all, this would be a potentially observable change for them -- as far as I can tell, even while in IBD we can at least sometimes validate and relay a transaction just fine, if a transaction were only spending utxos that are already confirmed...  (Unless I'm missing some other interaction that would prevent that from working?)

Not to say that this change can't be made, I'm just looking for reasoning and a comment that explains our actual thinking, because so far I think the reasoning has been contradictory.  If it's not a protocol violation to ignore unrequested transactions, I don't see the harm in us always doing it and avoiding gating yet another behavior on whether we are in IBD (which has always been a source of headaches when developers try to understand why something isn't working as expected).  If it is a protocol violation or potentially breaking, we should note that here and have a better justification for doing it than has been provided so far (particularly if a transaction were coming from a PF_RELAY or PF_FORCERELAY peer, where you might expect us to try harder to pass the transaction on).

Reply:
For some context, the comment is meant to convey that sending us unsolicited transactions is not a protocol violation, and I wanted to distinguish from the case above in which we disconnect the peer in addition to ignoring the transaction for violating no-tx-relay. Something like "since the peer has not violated protocol, don't punish" is what I meant. The comment could definitely be clearer.

My goal is just to fail early when we don't expect to be able to validate a transaction. It's definitely a good point that we could be perfectly able to validate despite being in IBD. I wonder if we could improve the condition with a better heuristic than "if in IBD?"

Reply:
Ah, yes I completely misparsed that protocol violation comment, thanks for explaining that.

I'm still looking to understand why this change is beneficial, but a broader change that includes IBD as a subset and also allows for PF_RELAY peers to bypass it, would not be.

Reply:
I think we can consider this change as beneficial as it's a small resources saving during IBD, a period during which our resources should be allocated on a more prioritized task. W.r.t to a broader change adopting this behavior beyond IBD (e.g #21224) we have an explicit concern about silently harming other bitcoin softwares deployed on the tx-relay network. As it's a reasonable assumption to expect only a minority fraction of nodes doing IBD at anytime, this change might be considered as a far lesser disruption.

Though, we could also say that disruption would be limited for a broader change as we expect deployment cycles to be gradual.

> I wonder if we could improve the condition with a better heuristic than "if in IBD?"

You can definitely validate while in IBD. But "already confirmed utxos" at your-IBD-tip doesn't mean those have not been spent at network tip. If your last synced block is 100,  your valid transaction might already have been double-spent at block 200, so validation cost would have been lost. IMO, utxo set view should be assumed to fluctuate quickly during IBD. Also, you won't request transactions so your mempool is likely to be empty and you won't be able to accurately weight transaction feerate. Relying on those _heuristics_, better to save on resources until your node is more able to assert with certainty if the transaction is a good candidate for propagation and block inclusion.

Reply:
>Ah, yes I completely misparsed that protocol violation comment, thanks for explaining that.

I just had the same confusion.